### PR TITLE
Web Inspector Page Agent

### PR DESCRIPTION
**Pull request by KristinaKoeva**

### CODE REVIEW

File: src/NativeScript/TNSRuntime.mm [hunk 1]
```diff
@@ -50,28 +51,36 @@ - (JSGlobalContextRef)globalContext {
...
+        return exec->vm().throwException(exec, exception);
+    }
+
+    return JSFunction::create(exec->vm(), function, globalObject);
+}
+
 static JSC_HOST_CALL EncodedJSValue createModuleFunction(ExecState* execState) {
-    JSString* moduleBody = jsString(execState, execState->argument(0).toWTFString(execState));
+    WTF::String moduleBody = execState->argument(0).toWTFString(execState);
     WTF::String moduleUrl = execState->argument(1).toString(execState)->value(execState);
```
Comment (fealebenpae) [inline src/NativeScript/TNSRuntime.mm]: Update this to use `JSValue::toWTFString` as well.

File: src/NativeScript/TNSRuntime.mm [hunk 2]
```diff
@@ -50,28 +51,36 @@ - (JSGlobalContextRef)globalContext {
     return toGlobalRef(self->_globalObject->globalExec());
 }
 
+JSObject* constructFunction(
```
Comment (jasssonpet) [inline src/NativeScript/TNSRuntime.mm]: Can this function be made static?

File: src/NativeScript/inspector/CachedResource.mm [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    if (mimeType.startsWith("font/")) {
+        return Inspector::Protocol::Page::ResourceType::Font;
+    }
+
+    return Inspector::Protocol::Page::ResourceType::Other;
+}
+
+WTF::String CachedResource::content(ErrorString& out_error) {
+    if (m_content.isEmpty()) {
+        NSString* cachedResourcePath = (NSString*)m_url;
```
Comment (fealebenpae) [inline src/NativeScript/inspector/CachedResource.mm]: Use `static_cast`, `reinterpret_cast` or `const_cast` where possible, refrain at all from C-style casts because they can strip `const` qualifiers and are potentially unsafe.

File: src/NativeScript/inspector/CachedResource.mm [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+CachedResource::CachedResource() {}
+
+CachedResource::CachedResource(WTF::String url)
+    : m_content(WTF::emptyString()) {
+    m_url = url;
+
+    NSString* cachedResourcePath = (NSString*)url;
+    NSURL* cachedResourceUrl = [NSURL URLWithString:cachedResourcePath];
+
+    m_mimeType = WTF::String(NativeScript::mimeTypeByExtension([cachedResourceUrl pathExtension]));
```
Comment (fealebenpae) [inline src/NativeScript/inspector/CachedResource.mm]: If you only need `NSURL` for `pathExtension` you can use `-[NSString pathExtension]` directly.

File: src/NativeScript/inspector/GlobalObjectDebuggerAgent.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    m_globalObject = jsCast<NativeScript::GlobalObject*>(&globalObject);
+}
+
+void GlobalObjectDebuggerAgent::enable(ErrorString& errorString) {
+    InspectorDebuggerAgent::enable(errorString);
+
+    ResourceManager& resourceManager = ResourceManager::getInstance();
+    WTF::HashMap<WTF::String, WTF::RefPtr<JSC::SourceProvider>>& sourceProviders = resourceManager.sourceProviders();
+
+    for (auto iterator = sourceProviders.begin(); iterator != sourceProviders.end(); ++iterator) {
```
Comment (fealebenpae) [inline src/NativeScript/inspector/GlobalObjectDebuggerAgent.cpp]: If you like, you can use [range-based for loops](http://en.cppreference.com/w/cpp/language/range-for) because they are prettier than plain iterators.

File: src/NativeScript/inspector/GlobalObjectDebuggerAgent.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef GlobalObjectDebuggerAgent_h
+#define GlobalObjectDebuggerAgent_h
+
+#include "InspectorDebuggerAgent.h"
```
Comment (fealebenpae) [inline src/NativeScript/inspector/GlobalObjectDebuggerAgent.h]: There are includes of JavaScriptCore headers, right? Prefer using angle brackets:
`#include <JavaScriptCore/InspectorDebuggerAgent.h>`
where possible to make it easier to distinguish between third-party headers and our headers.

**Comment by KristinaKoeva** [comment]
ping @fealebenpae @jasssonpet your feedback has been applied. 
As per @fealebenpae  suggestion another commit was added that strips the bundle path from the resource url and leaves only file:///app/ as prefix.

**Comment by KristinaKoeva** [comment]
run ci

**Comment by KristinaKoeva** [comment]
run ci

File: src/NativeScript/inspector/MimeTypeHelper.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#ifndef NativeScript_MimeTypeHelper_h
+#define NativeScript_MimeTypeHelper_h
+
+#import <MobileCoreServices/MobileCoreServices.h>
+
+namespace NativeScript {
+
+static WTF::String mimeTypeByExtension(WTF::String extension) {
+    // UTI for iOS doesn't recognize css extensions and returns a dynamic UTI without a Mime Type
+    if(WTF::equal(extension, "css")) {
```
Comment (jasssonpet) [inline src/NativeScript/inspector/MimeTypeHelper.h]: Clang format?

**Comment by jasssonpet** [comment]
It seems to crash when a MIME type can not be found. It happens with the LICENSE file from the hello world template.

File: src/NativeScript/TNSRuntime.mm [hunk 8]
```diff
@@ -69,28 +70,36 @@ - (void)_onMemoryWarning {
...
+static JSObject* constructFunction(
+    ExecState* exec, JSGlobalObject* globalObject,
+    const Identifier& functionName, const String& sourceURL,
+    const String& moduleBody, const TextPosition& position) {
+    if (!globalObject->evalEnabled())
+        return exec->vm().throwException(exec, createEvalError(exec, globalObject->evalDisabledErrorMessage()));
+
+    ResourceManager& resourceManager = ResourceManager::getInstance();
+    WTF::RefPtr<SourceProvider> sourceProvider = resourceManager.addSourceProvider(sourceURL, moduleBody);
+    SourceCode source = SourceCode(sourceProvider, position.m_line.oneBasedInt(), position.m_column.oneBasedInt());
```
Comment (fealebenpae) [inline src/NativeScript/TNSRuntime.mm]: Use the inline constructor:
```
SourceCode source(sourceProvider, ..., ...);
```
in order to avoid the assignment operator

File: src/NativeScript/inspector/MimeTypeHelper.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace NativeScript {
+
+static WTF::String mimeTypeByExtension(WTF::String extension) {
+    if (extension.isEmpty()) {
+        return WTF::emptyString();
+    }
+
+    // UTI for iOS doesn't recognize css extensions and returns a dynamic UTI without a Mime Type
+    if (WTF::equal(extension, "css")) {
```
Comment (fealebenpae) [inline src/NativeScript/inspector/MimeTypeHelper.h]: #if PLATFORM(IOS)?

File: src/debugging/WebInspectorUI/Main.html [hunk 10]
```diff
@@ -487,6 +487,7 @@
     <script src="Views/Resizer.js"></script>
     <script src="Views/ResourceClusterContentView.js"></script>
     <script src="Views/ResourceDetailsSidebarPanel.js"></script>
+    <script src="Views/FileSystemResourcesHelper.js"></script>
```
Comment (fealebenpae) [inline src/debugging/WebInspectorUI/Main.html]: Is this the most correct name for this?

File: src/debugging/WebInspectorUI/Views/FileSystemResourcesHelper.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        this.resources = mainFrame.resources;
+        this.foldersByPath = new Map;
+        this.foldersByPath.set("/app", this);
+
+        this.expanded = true;
+    }
+
+    populate() 
+    {
+        for(var i = 0; i < this.resources.length; ++i)
```
Comment (fealebenpae) [inline src/debugging/WebInspectorUI/Views/FileSystemResourcesHelper.js]: for..of?

File: src/debugging/WebInspectorUI/Views/FileSystemResourcesHelper.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        return this._compareTreeElementsByMainTitle(a, b);
+    }
+
+    _generateParentFolderElement(pathComponents) 
+    {
+        var currentPath = "";
+        var nextPath = "";
+        var parentFolderElement;
+        for(var i = 0; i < pathComponents.length; i++) {
```
Comment (fealebenpae) [inline src/debugging/WebInspectorUI/Views/FileSystemResourcesHelper.js]: for..of?

File: src/debugging/debugger-proxy.js [hunk 13]
```diff
@@ -42,7 +42,7 @@ server.on("connection", function(webSocket) {
 	deviceSocket.pipe(packets);
 	
 	packets.on("data", function (buffer) {
-		//console.log("DEVICE " + buffer.toString("utf8"));
+		// console.log("DEVICE " + buffer.toString("utf8"));
```
Comment (fealebenpae) [inline src/debugging/debugger-proxy.js]: Is this change necessary? :)

File: src/debugging/debugger-proxy.js [hunk 14]
```diff
@@ -52,7 +52,7 @@ server.on("connection", function(webSocket) {
 	});
 	
 	webSocket.on("message", function (message, flags) {
-		//console.log("FRONTEND " + message);
+		// console.log("FRONTEND " + message);
```
Comment (fealebenpae) [inline src/debugging/debugger-proxy.js]: Ditto
